src/back/hlsl/help.rs

```diff
@@ -478,6 +837,46 @@ impl<'a, W: Write> super::Writer<'a, W> {
                         self.wrapped.constructors.insert(constructor);
                     }
                 }
+                // We treat matrices of the form matCx2 as a sequence of C vec2's due to
+                // differences in alignment between WGSL and HLSL for uniform buffers.
+                //
+                // The functions injected here are required to get the matrix accesses working.
+                crate::Expression::AccessIndex { base, index } => {
+                    let base_ty_res = &func_ctx.info[base].ty;
+                    let mut resolved = base_ty_res.inner_with(&module.types);
+                    let base_ty_handle = match *resolved {
+                        crate::TypeInner::Pointer { base, .. } => {
+                            resolved = &module.types[base].inner;
+                            Some(base)
+                        }
+                        _ => base_ty_res.handle(),
+                    };
+                    if let crate::TypeInner::Struct { ref members, .. } = *resolved {
+                        let member = &members[index as usize];
+
+                        match module.types[member.ty].inner {
+                            crate::TypeInner::Matrix { rows, .. }
+                                if member.binding.is_none() && rows == crate::VectorSize::Bi =>
+                            {
+                                let ty = base_ty_handle.unwrap();
+                                let access = WrappedStructMatrixAccess { ty, index };
+
+                                if !self.wrapped.struct_matrix_access.contains(&access) {
+                                    self.write_wrapped_struct_matrix_get_function(module, access)?;
+                                    self.write_wrapped_struct_matrix_set_function(module, access)?;
```

Comment:
I suspect this is right but still would like to hear your thoughts on one question. Since the transformation this PR is doing is really only needed for matCx2 in uniform buffers, why do we care about `set_xxx` functions? Uniform buffers are read-only after all.